packages/core/src/odata/common/request-builder/batch/batch-response-parser.ts

```diff
@@ -0,0 +1,376 @@
+      );
+    }
+  }
+  return {};
+}
+
+/**
+ * Parse a single response to the according response object type.
+ * @param response The string representation of a single response.
+ * @param entityToConstructorMap A map that holds the entity type to constructor mapping.
+ * @param responseDataAccessor Response data access module.
+ * @param deserializer Entity deserializer.
+ * @returns The parsed response in the according response object representation.
+ */
+export function parseResponse(
+  response: string,
+  entityToConstructorMap: Record<string, Constructable<EntityBase>>,
+  responseDataAccessor: ResponseDataAccessor,
+  deserializer: EntityDeserializer
+): WriteResponse | ReadResponse | ErrorResponse {
+  const responseData = {
+    body: parseResponseBody(response),
+    httpCode: parseHttpCode(response)
+  };
+
+  // retrieve response
+  if (responseData.httpCode === 200) {
+    return {
+      ...responseData,
+      type: getConstructor(
+        responseData.body,
+        entityToConstructorMap,
+        responseDataAccessor
+      ),
+      as: asReadResponse(responseData.body, responseDataAccessor, deserializer),
+      isSuccess: () => true
+    };
+  }
+
+  // change set sub response
+  if (responseData.httpCode === 201 || responseData.httpCode === 204) {
+    return {
+      ...responseData,
+      type: getConstructor(
+        responseData.body,
+        entityToConstructorMap,
+        responseDataAccessor
+      ),
+      as: asWriteResponse(responseData.body, responseDataAccessor, deserializer)
+    };
+  }
+
+  // error response
+  return { ...responseData, isSuccess: () => false };
+}
+
+/**
+ * Parse the complete batch HTTP response.
+ * @param batchResponse HTTP response.
+ * @param entityToConstructorMap A map that holds the entity type to constructor mapping.
+ * @param responseDataAccessor Response data access module.
+ * @param deserializer Entity deserializer.
+ * @returns An array of parsed sub responses of the batch response.
+ */
+export function parseBatchResponse(
+  batchResponse: HttpResponse,
+  entityToConstructorMap: Record<string, Constructable<EntityBase>>,
+  responseDataAccessor: ResponseDataAccessor,
+  deserializer: EntityDeserializer
+): (ErrorResponse | ReadResponse | WriteResponses)[] {
+  return splitBatchResponse(batchResponse).map(response => {
+    const contentType = getHeaderValue('content-type', parseHeaders(response));
+
+    // is change set response
+    if (isMultipartContentType(contentType)) {
+      return {
+        responses: splitChangeSetResponse(response).map(
+          subResponse =>
+            parseResponse(
+              subResponse,
+              entityToConstructorMap,
+              responseDataAccessor,
+              deserializer
+            ) as WriteResponse
+        ),
+        isSuccess: () => true
+      };
+    }
+
+    // is read or error response
+    if (isHttpContentType(contentType)) {
+      return parseResponse(
```

Comment:
this seems to be a bit unsymmetric. For change sets there is this subResponse and for ReadResponses not? But I guess this is correct.

Reply:
yes, that is how batch requests are built `{ batch: { retrieve, changeset: { change } } }`